File: frontend/run-window.js
/* Pure source-clock arithmetic. Date's UTC methods avoid host timezone/DST shifts. */
(function (root) {
  "use strict";
  const keys = ["sourceYear", "timeConvention", "startSeconds", "endSeconds"];
  const recipe = (window) =>
    window ? Object.fromEntries(keys.map((key) => [key, window[key]])) : null;
  function dateTime(year, seconds) {
    return new Date(Date.UTC(year, 0, 1) + seconds * 1000).toISOString().slice(0, 16);
  }
  function secondsFromDate(year, value) {
    if (!/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}$/.test(value)) return NaN;
    const time = Date.parse(`${value}:00Z`);
    if (!Number.isFinite(time) || new Date(time).toISOString().slice(0, 16) !== value) return NaN;
    return (time - Date.UTC(year, 0, 1)) / 1000;
  }
  function check(yearText, convention, startText, endText, bounds) {
    const year = Number(yearText);
    if (!Number.isInteger(year) || year < 1900 || year > 2100) return { error: "windowYearError" };
    if (!["source-local-standard", "UTC"].includes(convention))
      return { error: "windowConventionError" };
    const yearSeconds = (Date.UTC(year + 1, 0, 1) - Date.UTC(year, 0, 1)) / 1000;
    if (bounds.sourceEndSeconds > yearSeconds) return { error: "windowYearCoverageError" };
    const start = secondsFromDate(year, startText),
      end = secondsFromDate(year, endText);
    if (![start, end].every((value) => Number.isInteger(value) && value % 900 === 0))
      return { error: "windowGridError" };
    if (start >= end) return { error: "windowOrderError" };
    if (start < bounds.earliestStartSeconds || end > bounds.latestEndSeconds)
      return { error: "windowCoverageError" };
    return {
      window: {
        sourceYear: year,
        timeConvention: convention,
        startSeconds: start,
        endSeconds: end,
      },
      steps: (end - start) / 900,
      hours: (end - start) / 3600,
    };
  }
  const api = { recipe, dateTime, secondsFromDate, check };
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  else root.GreenlightRunWindow = api;
})(typeof globalThis === "undefined" ? this : globalThis);

File: backend/run_window.py
"""Source-calendar window rules, independent of the optional scientific runtime."""

from __future__ import annotations

import calendar
from datetime import datetime, timedelta
from typing import Any, Mapping

STEP_SECONDS = 900
DAY_SECONDS = 86400
LOOKAHEAD_SECONDS = 43200


def coverage(source_start: int, source_end: int) -> dict[str, int]:
    # The upstream lamp controller needs the daily radiation sum. Only offer
    # complete source days; no missing midnight context is treated as zero.
    earliest = ((source_start + DAY_SECONDS - 1) // DAY_SECONDS) * DAY_SECONDS
    latest = min(
        (source_end - LOOKAHEAD_SECONDS) // STEP_SECONDS * STEP_SECONDS,
        source_end // DAY_SECONDS * DAY_SECONDS,
    )
    return {
        "sourceStartSeconds": source_start,
        "sourceEndSeconds": source_end,
        "earliestStartSeconds": earliest,
        "latestEndSeconds": latest,
        "stepSeconds": STEP_SECONDS,
        "lookaheadSeconds": LOOKAHEAD_SECONDS,
        "warmupSeconds": 0,
    }


def validate_window(request: Any, bounds: Mapping[str, int]) -> dict[str, Any]:
    if not isinstance(request, Mapping):
        raise ValueError(
            "Choose a simulation window and specify the source year and time convention."
        )
    required = {"sourceYear", "timeConvention", "startSeconds", "endSeconds"}
    if set(request) != required:
        raise ValueError(
            "runWindow requires sourceYear, timeConvention, startSeconds and endSeconds only."
        )
    year = request["sourceYear"]
    if type(year) is not int or not 1900 <= year <= 2100:
        raise ValueError("sourceYear must be an integer from 1900 to 2100.")
    convention = request["timeConvention"]
    if convention not in ("source-local-standard", "UTC"):
        raise ValueError(
            "Choose source-local-standard or UTC. Daylight-saving clock changes are not supported."
        )
    start, end = request["startSeconds"], request["endSeconds"]
    if any(type(value) is not int or value % STEP_SECONDS for value in (start, end)):
        raise ValueError("Start and end must align to 15-minute source-clock boundaries.")
    if start >= end:
        raise ValueError("End must be after start by at least 15 minutes.")
    year_seconds = (366 if calendar.isleap(year) else 365) * DAY_SECONDS
    if bounds["sourceEndSeconds"] > year_seconds or start < 0 or end > year_seconds:
        raise ValueError(
            "Weather timestamps exceed the selected source year; check the year and leap-day data."
        )
    if start < bounds["earliestStartSeconds"] or end > bounds["latestEndSeconds"]:
        raise ValueError(
            "The selected window needs uploaded weather from its first midnight through its last day, plus 12 hours after the end. No weather is extrapolated."
        )
    return {
        **dict(request),
        "stepSeconds": STEP_SECONDS,
        "totalSteps": (end - start) // STEP_SECONDS,
        "durationMinutes": (end - start) // 60,
        "lookaheadSeconds": LOOKAHEAD_SECONDS,
        "warmupSeconds": 0,
        "initialization": "GreenLight default states at the selected start; no warm-up",
        "startDateTime": source_datetime(year, start),
        "endDateTime": source_datetime(year, end),
    }


def source_datetime(year: int, seconds: int) -> str:
    """Naive source-clock string: never apply the host's timezone or DST."""
    return (datetime(year, 1, 1) + timedelta(seconds=seconds)).isoformat(timespec="minutes")
